20231212-modular-models.md

```diff
@@ -0,0 +1,229 @@
+
+## Proposed Syntax
+
+Below there's an example of how the models would look like with the proposed syntax.
+
+`core.fga`
+```
+module core
+
+type user
+type organization
+   define member : [user]
+   define admin : [user]
+type group
+    define member : [user]
+```
+
+`confluence.fga`
+```
+module confluence
+import "core.fga"
+
+extend type base.organization
+  relations
+    define can_create_space : [core.user] or admin
+    
+type space
+  relations
+    define organization : [core.organization] 
+    
+type page
+  relations
+    define space : [space]
+    define owner : [core.user]
+```
+
+`jira.fga`
+```
+module jira
+import "core.fga"
+
+extend type core.organization
+  relations
+    define can_create_project : [core.user] or admin
+
+type project
+  relations
+    define organization : [core.organization] 
+    
+type ticket
+  relations
+    define project : [project]
+    define owner : [core.user]
+```
+
+## Proposed Implementation
+
+The simplest way to implement this is to do it at the language + CLI level. 
+
+For example, with the models above, the CLI will combine the files and generate a model like the one below:
+
+`model.fga`
+```
+type core.user
+type core.organization
+   define member : [core.user]
+   define admin : [core.user]
+   define can_create_space : [core.user] or admin
+   define can_create_project : [core.user] or admin
+
+type core.group
+    define member : [core.user]
+    
+type confluence.space
+  relations
+    define organization : [core.organization] 
+    
+type confluence.page
+  relations
+    define space : [confluence.space]
+    define owner : [core.user]
+
+type jira.project
+  relations
+    define organization : [core.organization] 
+    
+type jira.ticket
+  relations
+    define project : [jira.project]
+    define owner : [core.user]
+```
+
+This implies check/write operations need to use the fully qualified type names. 
+
+For example, to check if the user `anne` user is the owner of the `engineering` space, the check would be:
+
+`check` (
+  `user`: `"core.user:anne"`, 
+  `relation`: `"owner"`, 
+  `object` : `"confluence.space:engineering"`)
+
+## Handling default modules
+
+The DSL will still support defining types without a module, e.g.
+
+```
+type user
+type organization
+   define member : [user]
+   define admin : [user]
+```
+
+If you want to reference these types from another module, you can use the `default` module name, which would be a protected keyword that can't be used to name your own modules.
+
+```
+type space
+  relations
+    define organization : [default.organization] 
+```
+
+This enables introducing modules in scenarios where you already have a production model that you can't easily change without impacting existing applications, as if you add existing types to a module, the types will be renamed to include the module name.
+
+## Open Questions
+
+- Should we allow adding relations to types from other modules? 
```

Comment:
> Should we allow adding relations to types from other modules? 

That's something we would definitively be interested in. We have differents teams working on (and owning) different domains of our app. So we would like to separate the model definition according to these domains. Even for a given type, relations are domain related so it would make sense for us to put them in their dedicated module


the following module is maintained by the core team:

```
module core

type user

type organization
  define member : [user]
  define admin : [user]

```


the following module is maintained by the payments team:
```
module payments
import "core.fga"

extends type core.organization
  define can_create_payment : [core.user] or admin
```

the following module is maintained by the invoices team:
```
module invoices
import "core.fga"

extends type core.organization
  define can_mark_invoice_as_verified : [core.user] or admin
```

Reply:
@aaguiarz 

First question: 
Absolutely yes. Some users (not admin) can be given that permission by a direct assignment.

Second question:
That is a very good question and indeed I would say that, as the `can_create_payment` relation belong to the payments domain, the domain team is the owner of that information and so it should be the only one allowed to "manage" (write/delete) related tuples.

Who should write/delete which tuple/relation is a subject we currently are studying before implementing more use cases. We already draft some ideas, this will probably go further the scope of this rfc but perhaps it will give you ideas so i'll share it here.
In our case we are taking advantage that we implemented a facade service which is the only one to "speak" with openfga.
All other business services speak to the facade using openid authentication.

For now (our model is still pretty small) only a single service manages the tuples (other services only perform checks). But we have a lot of pending use cases and we'd like other services to be able to perform tuples write/delete providing enough isolation and reduce security hazard (kind of like what you mentioned in your question).

So here are the conditions we are thinking about to implement in our facade.

First, only one service can call the facade with an access token coming from a client_credentials authentication flow. It is the one to setup the initial `administrator` relation on an `organization` (that's basically what we already have).

To all other cases, we believe that no application is allowed to give permissions. Only end users are (or not) according to their own permission(s).
To do so, all other service must provide 2 access tokens:
  * one comming from a client_credentials authentication flow (this is to ensure that the caller is an authorized service)
  * forward the user access token (coming from an authorization_code flow performed through our frontend) so we have the actual user id

When such a call is made to the facade, it will lookup for "control relations" based on the query information and using the user access token user id.
Those control relations would follow some convention like:
 * `can_<create|delete>_<relation>` if the relations are defined on the same type of the query relation
 * `can_<create|delete>_<relation>_<objectType>` if the relations are defined on the type of the object of the query

Here is an authorization model that should illustrate it:
```
type user

type organization
  relations
    // managed by the core service
    define administrator: [user]

    // here control relations are defined on the same type
    define entities_administrator: [user]
    define can_create_entities_administrator: administrator  // checked by the facade when a end user try to write an 'entities_administrator' relation on this organization
    define can_delete_entities_administrator: administrator  // checked by the facade when a end user try to delete an 'entities_administrator' relation on this organization

    // here are control relations for the relation 'organization' defined in the type 'entity'
    define can_create_organization_entity: entities_administrator  // checked by the facade when a end user try to write an 'organization' relation on an entity having this organization as object
    define can_delete_organization_entity: entities_administrator  // checked by the facade when a end user try to delete an 'organization' relation on an entity having this organization as object


type entity
  relations
    // control relations are defined in the object type organization
    define organization: [organization]

    define administrator: entities_administrator from organization

    // here control relations are defined on the same type
    define manager: [user]
    define can_create_manager: administrator  // checked by the facade when a end user try to write a 'manager' relation on this entity
    define can_delete_manager: administrator  // checked by the facade when a end user try to delete a 'manager' relation on this entity

    define can_do_stuff: manager
```

In this example:
* only the core service can define an organization administrator (basically when a user creates a new organization)
* only an organization administrator can appoint a user as entities administrator
* only an entities administrator can create entities within the organization
* an entities administrator of an organization is the administrator of the organization entities
* an entity administrator can appoint a user as entity manager
* an entity manager can do stuff on the entity

I don't known if all that makes sense to you (it might not be all interesting for the current discussion ^^). And as I said this is a draft (we don't know at which point this is scalable with a more complex authorization model).

Reply:
Thanks for all the context @le-yams!

My understanding is that you are using the "control relations" approach for user identities. E.g. "you need to be an administrator to create a a manager", and that makes sense. I see that as a relatively common pattern when implementing FGA. To create specific instances of resources in your app, and the corresponding tuples, you need to have permission to do it. 

What we are trying to figure out is how to manage that for application identities too. In the access token you usually have a claim for the end user (`sub`) and another for the authorized party (`azp`) which would usually be a client_id. 

In the Atlassian example, a user would want to delete a space and be authorized to do it, but the Jira applications shouldn't delete the tuple, only the Confluence one. In that case, even if the user is authorized, deleting the tuple would fail. 

Using your approach, it could be something like:

define can_delete_manager: [application] or administrator

The Facade can now check if `application:<azp>' can_delete_manager. 

In the case of the end user, it can be more tricky to do it fully from the facade, because at that point you are already writing the tuple, and it's likely that you want to stop the user earlier (e.g. when the DELETE manager/<id> API call is executed, you'll check if the  user can delete a manager).

Makes sense?